### DISCUSSION TITLE
Teams Middleware and Testing

### DISCUSSION THREAD

**Discussion by mojowill** [spatie/laravel-permission#1947] [Q&A] [answered]
I'm struggling with getting this to work as expected when using teams.

First up is the creation of custom middleware as detailed here https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions#working-with-teams-permissions

I have created the middleware but am not sure my implementation in `Kernel.php` is correct. I have added the middelware to `$middlewarePriority` as suggested but is it in the right place?

```php
protected $middlewarePriority = [
        \Illuminate\Cookie\Middleware\EncryptCookies::class,
        StartSession::class,
        ShareErrorsFromSession::class,
        AuthenticatesRequests::class,
        ThrottleRequests::class,
        ThrottleRequestsWithRedis::class,
        AuthenticateSession::class,
        SubstituteBindings::class,
        TeamsPermission::class, // my custom middleware
        Authorize::class,
    ];
```

It only seems to work however if I also register my middleware in the `$middlewareGroups`:

```php
protected $middlewareGroups = [
        'web' => [
            EncryptCookies::class,
            AddQueuedCookiesToResponse::class,
            StartSession::class,
            // \Illuminate\Session\Middleware\AuthenticateSession::class,
            ShareErrorsFromSession::class,
            VerifyCsrfToken::class,
            SubstituteBindings::class,
            TeamsPermission::class, // my custom middleware
            CheckIsSuspended::class
        ],
```

Any additional guidance would be appreciated.

Secondly testing.....

I am running my seeds during `SetUp()` however we need to test roles during some of our tests, generally in a test we use `setPermissionsTeamId()` to set to the appropriate team and then assign a role to the user, however this doesn't appear to work consistently or as expected, in some cases tests pass, in others they don't with 403 errors when not expected. When debugging I can see that the role is assigned but the `session('team_id')` (which is what we use to track and set teams) isn't always in-sync with `app(PermissionRegistrar::class)->getPermissionsTeamId()`.

I am assuming that perhaps my second point is related to my first and our middleware isn't correct?

Paging @erikn69 as he seems to be the legend when it comes to team permissions 😉

**Comment by erikn69** [comment]
> It only seems to work however if I also register my middleware in the $middlewareGroups:

Yes, it must be on `$middlewareGroups`(important), but you can make your `team_id` setter by other ways
`$middlewarePriority` is opcional, but when you use other middleware could be needed, here is mine: https://github.com/spatie/laravel-permission/discussions/1906#discussioncomment-1584335

----

>  in some cases tests pass, in others they don't with 403 errors when not expected

Maybe you aren't using tests correctly, but i can't know that
#1894 https://laravel.com/docs/8.x/database-testing#persisting-models
maybe `RefreshDatabase`?

**UPDATE:**
Also your middleware maybe don't work on testing, use  `setPermissionsTeamId()`, and before test `$user->load('roles','permissions');` for sync with the setted `team_id`, example:
https://github.com/spatie/laravel-permission/blob/08cc9bdbf89640cb1d64fb01eb9513c1cbb2549d/tests/TeamHasRolesTest.php#L25-L27

**Reply by mojowill** [reply]
Thanks it does point to my tests being issue, and I think specifically it is around trying to assignRoles for users during tests. What I am finding is if I run the same test multiple times and compare my `$team-id` `session('team_id')` and `app(PermissionRegistrar::class)->getPermissionsTeamId()` the `getPermissionsTeamId()` won't always be as expected.

I am using `LazilyRefreshDatabases` on my base Test Case, an example of one of my test:

```php
<?php

namespace Tests\Feature\Livewire\Folders;

use App\Http\Livewire\Folder\CreateFolder;
use App\Models\Team;
use App\Models\User;
use Livewire\Livewire;
use Spatie\Permission\PermissionRegistrar;
use Tests\TestCase;

class CreateTest extends TestCase
{
    private Team $team;

    private Team $team2;

    private User $user;

    private User $user2;

    protected function setUp(): void
    {
        parent::setUp();

        [$this->team, $this->team2] = Team::factory(2)->create();

        $this->user = User::factory()
            ->hasAttached($this->team)
            ->hasAttached($this->team2)
            ->create();

        $this->user2 = User::factory()
            ->hasAttached($this->team)
            ->create();
    }

    public function test_only_authorized_users_can_create_a_folder()
    {
        // Try to create a folder in a team we belong to and have the correct rights
        app(PermissionRegistrar::class)->setPermissionsTeamId($this->team->id);
        $this->user->assignRole('admin');

        $this->actingAs($this->user)
            ->withSession(['team_id' => $this->team->id])
            ->get(route('dashboard'));

        Livewire::test(CreateFolder::class, ['teamId' => $this->team->id])
            ->set('name', 'Test Folder')
            ->set('folderUsers', [$this->user->id, $this->user2->id])
            ->call('create')
            ->assertSuccessful();

        // Try to create a folder in a team we belong to but don't have the correct rights
        app(PermissionRegistrar::class)->setPermissionsTeamId($this->team2->id);

        $this->actingAs($this->user2)
            ->withSession(['team_id' => $this->team2->id])
            ->get(route('dashboard'));

        Livewire::test(CreateFolder::class, ['teamId' => $this->team->id])
            ->set('name', 'Test Folder')
            ->call('create')
            ->assertForbidden();
    }

    public function test_validation_works()
    {
        ray(app(PermissionRegistrar::class)->getPermissionsTeamId())->red();

        app(PermissionRegistrar::class)->setPermissionsTeamId($this->team->id);
        $this->user->assignRole('admin');

        $this->actingAs($this->user)
            ->withSession(['team_id' => $this->team->id])
            ->get(route('dashboard'));

        ray(['spatie-team' => app(PermissionRegistrar::class)->getPermissionsTeamId(), 'session-team' => session('team_id'), 'test-team' => $this->team->id ]);

        $response = Livewire::test(CreateFolder::class, ['teamId' => $this->team->id])
            ->set('name', 'Test Folder 2')
            ->set('folderUsers', [])
            ->call('create');

        ray(app(PermissionRegistrar::class)->getPermissionsTeamId())->orange();

        $response->assertSuccessful()
            ->assertHasErrors('folderUsers');

        ray(app(PermissionRegistrar::class)->getPermissionsTeamId())->green();

    }
}
```

It is always the `assertSuccessful()` that causes the issues, sometimes it works, sometimes it doesn't 😕

**Reply by erikn69** [reply]
> Also your middleware maybe don't work on testing, use setPermissionsTeamId(), and before test $user->load('roles','permissions'); for sync with the setted team_id, example:
https://github.com/spatie/laravel-permission/blob/08cc9bdbf89640cb1d64fb01eb9513c1cbb2549d/tests/TeamHasRolesTest.php#L25-L27

Try that
After `setPermissionsTeamId()` always reload relations on tests, for the new pivot

**Reply by mojowill** [reply]
Same result unfortunately every few runs the value of `getPermissionsTeamId()` seems to change for no reason

```php
public function test_validation_works()
    {
        setPermissionsTeamId($this->team->id);
        $this->user->load('roles');
        $this->user->assignRole('admin');
        
        $this->actingAs($this->user)
            ->withSession(['team_id' => $this->team->id])
            ->get(route('dashboard'));

        Livewire::test(CreateFolder::class, ['teamId' => $this->team->id])
            ->set('name', 'Test Folder 2')
            ->set('folderUsers', [])
            ->call('create')
            ->assertSuccessful()
            ->assertHasErrors('folderUsers');
    }
```

**Reply by erikn69** [reply]
i am puzzled, could you create a minimal Laravel app with the test that demonstrates the issue? [example](https://github.com/Onekone/spatie-issue-example/blob/main/tests/Feature/IssueTest.php)

**Reply by mojowill** [reply]
I don't think I can create a minimal laravel app due to the number of internal dependancies this app has on itself. Happy to grant you read access to the repo though?

**Reply by erikn69** [reply]
>due to the number of internal dependancies this app

The idea is to isolate permissions from that, many times when using package on a fresh installation found that the package was not the problem

**Reply by erikn69** [reply]
when you create the role `admin`, do you set `team_id` to null?

**Reply by mojowill** [reply]
Yes all roles in the seeder are set with team_id null as they are global roles

**Reply by erikn69** [reply]
I did make a little test and i think i found the problem
```php
setPermissionsTeamId(1);
$user->unsetRelation('roles')->unsetRelation('permissions');
$this->actingAs($user)
    ->withSession(['team_id' => 3])
    ->get('tokens/test-success')
    ->assertForbidden();
```
Using `unsetRelation` lets the middleware do his work

Other tip, did you add `team_id` to recreate session on remember-me functionality?

**Reply by mojowill** [reply]
will test the unsetRelation as for the team_id no I haven't added it elsewhere/.... 😖

**Reply by erikn69** [reply]
> Ok thanks that works but I experiencing the same issues as the other tests the team id and getPermissionsId do not always match?

everything checked, i can't replicate your issue, i can't do more
```php
/** This is permitted */
public function test_WebSuccess()
{
    $user = $this->prepare();
    setPermissionsTeamId(2);
    $user->unsetRelation('roles')->unsetRelation('permissions');
    $this->actingAs($user)
        ->withSession(['team_id' => 1])
        ->get('tokens/test-success')
        ->assertOk();
    $this->assertEquals(1, app(PermissionRegistrar::class)->getPermissionsTeamId());
}

/** This is forbidden */
public function test_WebFailed()
{
    $user = $this->prepare();        
    setPermissionsTeamId(1);
    $user->unsetRelation('roles')->unsetRelation('permissions');
    $this->actingAs($user)
        ->withSession(['team_id' => 2])
        ->get('tokens/test-success')
        ->assertForbidden();
    $this->assertEquals(2, app(PermissionRegistrar::class)->getPermissionsTeamId());
}
```
```batch
PHPUnit 9.5.10 by Sebastian Bergmann and contributors.
..                                                                  2 / 2 (100%)
Time: 00:00.326, Memory: 28.00 MB
OK ( 2 tests, 14 assertions)
```

**Comment by mojowill** [comment]
Perhaps I'm being dumb, which is entirely possible. But getPermissionsTeamId doesn't appear to be working for me as I would expect?

I created a simple test to just test roles and permissions were working as I expected without the rest of the app following the example you provided, when I add a test to check the value of getPermissionsTeamId it comes back as null?

```php
public function test_that_the_team_id_is_set_correctly_via_the_middleware()
    {
        setPermissionsTeamId($this->team->id);
        $this->user2->load('roles');
        $this->user->assignRole('admin', 'manager');
        
        $this->actingAs($this->user)
            ->withSession(['team_id' => $this->team->id])
            ->get(route('dashboard'));

        $this->assertEquals($this->team->id, getPermissionsTeamId());
    }
```

**Reply by erikn69** [reply]
that is another problem to be resolved #1944

>[IMPORTANT] Fixed missing return on getPermissionsTeamId() helper
https://github.com/spatie/laravel-permission/blob/1c9304af746b42308a1d1da2989e12a1439571a9/src/helpers.php#L37-L40

use `app(PermissionRegistrar::class)->getPermissionsTeamId()` for the moment, authors not responding

**Reply by mojowill** [reply]
Ok thanks that works but I experiencing the same issues as the other tests the team id and getPermissionsId do not always match?
![Screenshot 2021-12-02 at 16 02 56](https://user-images.githubusercontent.com/865606/144458157-c5b5e0a1-eaf6-44bb-9fdd-be2c08daaa46.png)
?

**Reply by erikn69** [reply]
```php
setPermissionsTeamId($this->team);
```
You can set `$this->team` directly, not `->id` needed

**Reply by mojowill** [reply]
So I think i fixed the issue in my Permissions test (I was getting the roles for a different team in my dashboard and not resetting setPermissionsTeamId back to what it should be)

However my original test is still showing the same issues where sometimes it passes and sometimes it fails

**Comment by mojowill** [comment, marked as answer]
I think my main issue is how I've implemented a method to return the users roles for a given team (ie what roles have been assigned to the user for the given team).

Is there a cleaner way to do this?

```php
/**
     * Returns the Roles associated with the current user for the given team
     *
     * @param  User  $user
     * @param  int  $teamId
     * @return Collection|mixed|Role[]
     */
    public function getTeamRoles(self $user, int $teamId)
    {
        $originalTeam = app(PermissionRegistrar::class)->getPermissionsTeamId();
        app(PermissionRegistrar::class)->setPermissionsTeamId($teamId);

        $user->load('roles');

        setPermissionsTeamId($originalTeam);

        return $user->roles;
    }
```

**Reply by erikn69** [reply]
```php
return Role::whereHas('users', function ($query) use ($user, $teamId) {
    $query->where(config('permission.column_names.model_morph_key'), $user->getKey())
        ->where(PermissionRegistrar::$teamsKey, $teamId);
})->get();
```

**Reply by mojowill** [reply]
Pure gold thanks! Just like that the issues are resolved!

**Reply by erikn69** [reply]
#1948
